csharp/src/Apache.Arrow/DecimalUtility.cs

```diff
@@ -0,0 +1,163 @@
+            {
+                BigInteger scaleBy = BigInteger.Pow(10, scale);
+                BigInteger integerPart = BigInteger.DivRem(integerValue, scaleBy, out BigInteger fractionalPart);
+                if (integerPart > _maxDecimal || integerPart < _minDecimal) // decimal overflow, not much we can do here - C# needs a BigDecimal
+                {
+                    throw new OverflowException("Value: " + integerPart + " too big or too small to be represented as a decimal");
+                }
+                return (decimal)integerPart + DivideByScale(fractionalPart, scale);
+            }
+            else
+            {
+                return DivideByScale(integerValue, scale);
+            }
+        }
+
+        private static decimal DivideByScale(BigInteger integerValue, int scale)
+        {
+            decimal result = (decimal)integerValue; // this cast is safe here
+            int drop = scale;
+            while (drop > PowersOfTenLength)
+            {
+                result /= s_powersOfTen[PowersOfTenLength];
+                drop -= PowersOfTenLength;
+            }
+
+            result /= s_powersOfTen[drop];
+            return result;
+        }
+
+        public static void GetBytes(BigInteger integerValue, int byteWidth, ref Span<byte> bytes)
+        {
+            if (bytes.Length != byteWidth)
+            {
+                throw new OverflowException("ValueBuffer size not equal to " + byteWidth + " byte width: " + bytes.Length);
+            }
+
+            Span<byte> integerBytes = integerValue.ToByteArray().AsSpan();
+            if (integerBytes.Length > byteWidth)
+            {
+                throw new OverflowException("Decimal size greater than " + byteWidth + " bytes: " + integerBytes.Length);
+            }
+
+            if (integerBytes.Length == byteWidth)
+            {
+                bytes = integerBytes;
+                return;
+            }
+
+            if (integerValue.Sign == -1)
+            {
+                integerBytes.CopyTo(bytes);
+                for (int i = integerBytes.Length; i < byteWidth; i++)
+                {
+                    bytes[i] = 255;
+                }
+            }
+            else
+            {
+                integerBytes.CopyTo(bytes);
+            }
+        }
+
+        public static bool CheckPrecisionAndScale(decimal value, int precision, int scale, out BigInteger integerValue)
+        {
+            DecimalLayout layout = new DecimalLayout(value); // use in place of decimal.GetBits(value) to avoid an allocation
+            integerValue = new BigInteger(BitConverter.GetBytes(layout.Lo).Concat(BitConverter.GetBytes(layout.Mid)).Concat(BitConverter.GetBytes(layout.Hi)).ToArray());
+
+            if (layout.Scale > scale)
+                throw new OverflowException("Decimal scale can not be greater than that in the Arrow vector: " + layout.Scale + " != " + scale);
```

Comment:
Do we have tests for these overflow scenarios?

Reply:
Added